# Regularization

##### Regularization is used to find if the model is overfitting or not.

#### Type:
1)L1-->Ridge
2)L2-->Lasso

In [6]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import Lasso, LassoCV, Ridge, RidgeCV
from sklearn.datasets import load_iris
from sklearn.metrics import r2_score

In [17]:
iris=load_iris()


In [19]:
iris.keys()

dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names', 'filename'])

In [21]:
iris.target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [22]:
x=iris.data
y=iris.target

In [23]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=.33,random_state=40)

In [24]:
lassocv=LassoCV(alphas=None, cv=10, max_iter=100, normalize=True)
lassocv.fit(x_train,y_train)

LassoCV(alphas=None, copy_X=True, cv=10, eps=0.001, fit_intercept=True,
        max_iter=100, n_alphas=100, n_jobs=None, normalize=True, positive=False,
        precompute='auto', random_state=None, selection='cyclic', tol=0.0001,
        verbose=False)

In [25]:
alpha=lassocv.alpha_
alpha

7.6501968706293e-05

In [26]:
lasso_reg=Lasso(alpha)
lasso_reg

Lasso(alpha=7.6501968706293e-05, copy_X=True, fit_intercept=True, max_iter=1000,
      normalize=False, positive=False, precompute=False, random_state=None,
      selection='cyclic', tol=0.0001, warm_start=False)

In [27]:
lasso_reg.fit(x_train,y_train)

Lasso(alpha=7.6501968706293e-05, copy_X=True, fit_intercept=True, max_iter=1000,
      normalize=False, positive=False, precompute=False, random_state=None,
      selection='cyclic', tol=0.0001, warm_start=False)

In [29]:
lasso_reg.score(x_test,y_test)

0.9425805954323527

In [37]:
alphas=np.random.uniform(low=0, high=10, size=(50,))
ridgecv=RidgeCV(alphas=alphas, cv=10, normalize=True)
ridgecv.fit(x_train,y_train)


RidgeCV(alphas=array([0.05453106, 8.42165984, 4.16580433, 2.64953345, 4.61146597,
       9.29707242, 1.59692663, 0.03032318, 0.55119253, 0.22992654,
       7.70636255, 1.61562059, 7.85952446, 2.28670675, 3.70485604,
       4.82028046, 8.77561759, 5.57867619, 2.7662038 , 3.19993232,
       8.62083078, 3.30539176, 4.33090916, 1.38001806, 6.64605161,
       2.22598439, 1.09212681, 3.96095015, 2.06667113, 4.23213349,
       1.45078429, 6.6131633 , 2.85328353, 7.65026743, 7.04651983,
       0.94887661, 5.98295358, 7.90893968, 8.10457506, 0.17441963,
       7.8659192 , 7.54658247, 6.03651793, 7.14513317, 0.62143863,
       4.76977821, 0.14316017, 6.66909256, 9.18600549, 1.85427744]),
        cv=10, fit_intercept=True, gcv_mode=None, normalize=True, scoring=None,
        store_cv_values=False)

In [38]:
ridgecv.alpha_

0.030323180456677257

In [39]:
ridge_model=Ridge(alpha=ridgecv.alpha_)
ridge_model

Ridge(alpha=0.030323180456677257, copy_X=True, fit_intercept=True,
      max_iter=None, normalize=False, random_state=None, solver='auto',
      tol=0.001)

In [40]:
ridge_model.fit(x_train, y_train)

Ridge(alpha=0.030323180456677257, copy_X=True, fit_intercept=True,
      max_iter=None, normalize=False, random_state=None, solver='auto',
      tol=0.001)

In [41]:
ridge_model.score(x_test,y_test)

0.9425435171686609

###### compare the 2 model and choose the best model. In this model Lasso performed well 0.9425805954323527